src/Sarif/Baseline/SarifLogBaseliner.cs

```diff
@@ -42,5 +48,68 @@ public Run CreateBaselinedRun(Run baseLine, Run nextLog)
 
             return differencedRun;
         }
+
+        // Get the execution time from the specified run. This time will be used
+        // as the "first detection time" for results that don't already have that
+        // information. If the run doesn't have execution time information, it
+        // falls back to "now".
+        internal static DateTime GetRunTime(Run run, DateTime now)
+        {
+            DateTime runTime;
+
+            IList<Invocation> invocations = run.Invocations;
+            if (invocations?.Count > 0)
+            {
+                Invocation invocation = invocations[0];
+                runTime = invocation.EndTimeUtc;
+                if (runTime == DateTime.MinValue)
+                {
+                    runTime = invocation.StartTimeUtc;
+                }
+                if (runTime == DateTime.MinValue)
+                {
+                    runTime = now;
+                }
+            }
+            else
+            {
+                runTime = now;
+            }
+
+            return runTime;
+        }
+
+        // Set the baseline state and first detection time for the current result. If the
+        // matching result from the baseline has a first detection time, roll it forward.
+        // Otherwise, fall back to the time that the run was performed (for existing
+        // results, look at the baseline run; for new results, look at the current run).
+        // If the run doesn't have execution time information, it falls back to "now"
+        // (see GetRunTime).
+        internal static void SetResultBaselineInformation(
+            Result currentResult,
+            Result baselineResult, DateTime newRunTime, DateTime baselineRunTime)
+        {
+            if (baselineResult == null)
+            {
+                currentResult.BaselineState = BaselineState.New;
+                SetFirstDetectionTime(currentResult, newRunTime);
+            }
+            else
+            {
+                currentResult.BaselineState = BaselineState.Unchanged;
+
+                DateTime firstDetectionTime = baselineResult.Provenance != null && baselineResult.Provenance.FirstDetectionTimeUtc != DateTime.MinValue
```

Comment:
It might be clearer if you merge all of this into the method - see if a Provenance exists, if there's a previous FirstDetectionTime, if it's less than what you otherwise intend to set, and then set it.